### PR TITLE
reconcile services

### PR DESCRIPTION
**Pull request by calvix**
As the annotation is now needed for guest cluster monitoring we need to be able to update old cluster and the that means enable reconciliation of the service to desired state

Also since the etcd port is different across providers and it needs to be part of annotation

### CODE REVIEW

File: service/controller/v14/resource/service/master_service.go [hunk 1]
```diff
@@ -15,7 +15,8 @@ func newMasterService(customObject v1alpha1.KVMConfig) *apiv1.Service {
 			APIVersion: "v1",
 		},
 		ObjectMeta: apismetav1.ObjectMeta{
-			Name: key.MasterID,
+			Name:      key.MasterID,
+			Namespace: key.ClusterID(customObject),
```
Comment (calvix) [inline service/controller/v14/resource/service/master_service.go]: for some reason, ns was missing here all the time

**Comment by calvix** [comment]
double tested, all good and work as expected

File: service/controller/v14/resource/service/master_service.go (hunk 1 above)
Comment (MarcelMue) [reply service/controller/v14/resource/service/master_service.go]: What happens when we upgrade clusters now? Will there be leftovers in the old ns that was used?

File: service/controller/v14/resource/service/resource.go [hunk 2]
```diff
@@ -81,3 +112,26 @@ func toServices(v interface{}) ([]*apiv1.Service, error) {
 
 	return services, nil
 }
+
+// portsEqual is a function that is checking if ports in the service have same important values.
+func portsEqual(a, b *apiv1.Service) bool {
```
Comment (MarcelMue) [inline service/controller/v14/resource/service/resource.go]: I would like to see some unit tests for this one as well.

**Review by MarcelMue — COMMENTED** [review]
Seems pretty good over all, just a few questions.

File: service/controller/v14/resource/service/resource.go (hunk 2 above)
Comment (calvix) [reply service/controller/v14/resource/service/resource.go]: there is already test for it in the `isServiceModified` as this one is used there, does it make sens to copy the test case for this function?

File: service/controller/v14/resource/service/master_service.go (hunk 1 above)
Comment (calvix) [reply service/controller/v14/resource/service/master_service.go]: It wasn't really a big problem as  when you call to create a function you specify the ns int he calls
```
k8s.CoreV1().Services(namespace).Create(service)
```
but when I was comparing the current and desired service the NS was missing in the desired resource.

TL;DR: no leftovers, all will be same.

Comment (MarcelMue) [reply service/controller/v14/resource/service/master_service.go]: That's good then :+1:

File: service/controller/v14/resource/service/resource.go (hunk 2 above)
Comment (MarcelMue) [reply service/controller/v14/resource/service/resource.go]: Yes you have a test for the mismatch. The benefit of testing the individual function is if someone changes something in the function they won't have to search for the reason why a different test is failing. A unit test should make sure the function itself works as expected.

Additionally I don't believe that you have situations tested like one service doesn't have ports at all, there are no ports on either service and so on.

This is my opinion here, if you feel comfortable with this - go ahead.

**Review by MarcelMue — APPROVED** [review]
Fine with me if you feel comfortable with this.

**Review by xh3b4sd — CHANGES_REQUESTED** [review]
IMO we should refactor the status resource and make it usable for all providers regardless. The differences probably can be made configurable but 99% of all of this code is now duplicated across at least 3 repos.

File: service/controller/v14/resource/service/resource.go (hunk 2 above)
Comment (calvix) [reply service/controller/v14/resource/service/resource.go]: Yep, you were right, i added test fro that function too.

**Comment by calvix** [comment]
RFR

**Review by xh3b4sd — APPROVED** [review]
When this works and is tested lets go. We will refactor this in the SIG.

File: service/controller/v14/resource/service/update.go [hunk 3]
```diff
@@ -30,5 +51,41 @@ func (r *Resource) NewUpdatePatch(ctx context.Context, obj, currentState, desire
...
+			continue
+		} else if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		if isServiceModified(desiredService, currentService) {
+			// Make a copy and set the resource version so the service can be updated.
+			serviceToUpdate := desiredService.DeepCopy()
+			serviceToUpdate.ObjectMeta.ResourceVersion = currentService.ObjectMeta.ResourceVersion
+			serviceToUpdate.Spec.ClusterIP = currentService.Spec.ClusterIP
```
Comment (xh3b4sd) [inline service/controller/v14/resource/service/update.go]: I am hitting issues in KVM when updating from `2.10.0` to `2.11.0`. The update to k8s fails and complains the service is outdated. Did you test the update from one version to another? Why do we set the resource version here explicitly?

Comment (xh3b4sd) [reply service/controller/v14/resource/service/update.go]: Ok this is bullshit. I probably got confused with the statusresource magic I hustled with meanwhile. All good. Never mind. <3
